dartagnan/src/main/java/com/dat3m/dartagnan/utils/dependable/DependencyGraph.java

```diff
@@ -118,47 +121,87 @@ private void tarjan() {
 
         // Create sorted list of all nodes
         // and sort their dependencies/dependents as well
-        Comparator<Node> cmp = Comparator.comparingInt(Node::getTopologicalIndex);
+        final Comparator<Node> cmp = Comparator.comparingInt(Node::getTopologicalIndex);
         nodeList = ImmutableList.sortedCopyOf(cmp, nodeMap.values());
         for (Node node : nodeList){
             node.getDependencies().sort(cmp);
             node.getDependents().sort(cmp);
         }
 
         Collections.reverse(sccs);
-
     }
 
-    private void strongConnect(Node v) {
-        v.index = index;
-        v.lowlink = index;
-        stack.push(v);
-        v.isOnStack = true;
-        index++;
+    /*
+        A non-recursive implementation of the recursive DFS-based algorithm
+        found at the bottom of this class.
+     */
+    private void strongConnect(Deque<StackFrame> recursionStack) {
+        Node lastVisited = null;
+        while (!recursionStack.isEmpty()) {
+            final StackFrame frame = recursionStack.getLast();
+            final Node v = frame.node;
+            final Iterator<Node> successorsToProcess = frame.iterator;
+
+            if (!v.wasVisited()) {
+                // Start of recursion
+                v.index = index;
+                v.lowlink = index;
+                stack.push(v);
+                v.isOnStack = true;
+                index++;
+            } else if (lastVisited != null) {
+                // Recursion in progress (we returned from a deeper recursion level)
+                v.lowlink = Math.min(v.lowlink, lastVisited.lowlink);
+            } else {
+                assert false;
```

Comment:
I guess this point should never be reached. Maybe we should add an error msg to this sanity check.

Reply:
I don't think an error message is reasonable for these kinds of invariants. But I will change the code to just have a `else { assert (cond); ...}` rather than `else if (cond) { ...} else { assert false; }`

Reply:
Once you do this change, I can merge